testgrid/cmd/config/main.go

```diff
@@ -17,60 +17,201 @@ limitations under the License.
-	yamlPaths  = flag.String("yaml", "", "comma-separated list of input YAML files")
-	printText  = flag.Bool("print-text", false, "print generated proto in text format to stdout")
-	outputPath = flag.String("output", "", "output path to save generated protobuf data")
+	"cloud.google.com/go/storage"
 )
 
-func errExit(format string, a ...interface{}) {
-	fmt.Fprintf(os.Stderr, format, a...)
-	os.Exit(1)
+type multiString []string
+
+func (m multiString) String() string {
+	return strings.Join(m, ",")
 }
 
-func main() {
+func (m *multiString) Set(v string) error {
+	*m = strings.Split(v, ",")
+	return nil
+}
+
+type options struct {
+	creds     string
+	inputs    multiString
+	oneshot   bool
+	output    string
+	printText bool
+}
+
+func gatherOptions() options {
+	o := options{}
+	flag.StringVar(&o.creds, "gcp-service-account", "", "/path/to/gcp/creds (use local creds if empty")
+	flag.BoolVar(&o.oneshot, "oneshot", false, "Write proto once and exit instead of monitoring --yaml files for changes")
+	flag.StringVar(&o.output, "output", "", "write proto to gs://bucket/obj or /local/path")
+	flag.BoolVar(&o.printText, "print-text", false, "print generated proto in text format to stdout")
+	flag.Var(&o.inputs, "yaml", "comma-separated list of input YAML files")
 	flag.Parse()
+	return o
+}
 
-	yamlFiles := strings.Split(*yamlPaths, ",")
-	if len(yamlFiles) == 0 || yamlFiles[0] == "" {
-		errExit("Must specify one or more YAML files with --yaml\n")
+func (o *options) validate() error {
+	if len(o.inputs) == 0 || o.inputs[0] == "" {
+		return errors.New("--yaml must include at least one file")
 	}
-	if !*printText && *outputPath == "" {
-		errExit("Must set --print-text or --output\n")
+
+	if !o.printText && o.output == "" {
+		return errors.New("--print-text or --output=gs://path required")
 	}
-	if *printText && *outputPath != "" {
-		errExit("Cannot set both --print-text and --output\n")
+	return nil
+}
+
+// announceChanges watches for changes to files and writes one of them to the channel
+func announceChanges(ctx context.Context, paths []string, channel chan []string) {
+	defer close(channel)
+	modified := map[string]time.Time{}
+	for _, p := range paths {
+		modified[p] = time.Time{} // Never seen
+	}
+
+	// TODO(fejta): consider waiting for a notification rather than polling
+	// but performance isn't that big a deal here.
+	for {
+		var changed []string
+		for p, last := range modified {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			info, err := os.Stat(p)
+			switch {
+			case os.IsNotExist(err) && !last.IsZero():
+				// File deleted
+				modified[p] = time.Time{}
+				changed = append(changed, p)
+			case err != nil:
+				log.Printf("Error reading %s: %v", p, err)
+			default:
+				if t := info.ModTime(); t.After(last) {
+					changed = append(changed, p)
+					modified[p] = t
+				}
+			}
+		}
+		if len(changed) > 0 {
+			select {
+			case channel <- changed:
+			case <-ctx.Done():
+				return
+			}
+		}
+		select {
```

Comment:
Shouldn't this be `else { select {` ?

Reply:
No guarantee that <-ctx.Done() is still empty

Reply:
Done